typed/remove_test.go

```diff
@@ -0,0 +1,133 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemove(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		deep   bool
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: ``,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `null`,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   true,
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   false,
+			// TODO: this seems to behave differently with a non deduced schema: {"a": "value", "b":{"c":"value"}} which is the reason for this PR
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			deep:   true,
+			// TODO: is a null result expected (insted of a {})
+			expect: `{"a": "value", "b": null}`,
+		},
+		{
+			object: `{"a": "value", "b": []}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   true,
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			deep:   true,
+			expect: `{"a": "value", "b": null}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c", "d"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			deep:   true,
+			// TODO: is this expected?
```

Comment:
Certainly not? You're not supposed to remove `d` are you?

Reply:
Good thing we have tests :-)

Reply:
I am not sure. I thought maybe it's not possible to remove items of a list by their data. But then this behavior still would be unexpected.

Reply:
fieldpath.MakePathOrDie("b", "c") does this path refer to the list item "c"? I think the path being invalid might be why this is happening, but either way the output isn't what I'd expect.

Reply:
If you want to reference the list item by it's data you might have to do: fieldpath.MakePathOrDie("b", value.Value{StringValue: "c"}) or something like this which would work with the new value interface

Reply:
Okay, then I'll check tomorrow why an invalid path might remove the entire thing.
But yes, that was the goal. I just wanted to try of this works and stumbled into this, so it's a test now.

Reply:
@jennybuckley the reason is https://github.com/kubernetes-sigs/structured-merge-diff/pull/133/files#diff-11232f9aff4ebc184b767f3a3520214aR49 
because in this case the ElementRelationship in the List is atomic, so the function just returns and the remove ends up with a nil output for this atom.